File: rust_exercises/leetcode/Problem727/Solution.java
class Solution {
    public String minWindow(String S, String T) {
        int[][] dp = new int[S.length() + 1][T.length() + 1];
        int[][] start = new int[S.length() + 1][T.length() + 1];
        String result = null;
        
        for (int i = 1; i <= S.length(); i += 1) {
            for (int j = 1; j <= T.length(); j += 1) {
                if (S.charAt(i - 1) == T.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                    if (dp[i][j] > 1) {
                        start[i][j] = start[i - 1][j - 1];
                    } else {
                        start[i][j] = i;
                    }
                } else {
                    if (dp[i - 1][j] > dp[i][j - 1]) {
                        dp[i][j] = dp[i - 1][j];
                        start[i][j] = start[i - 1][j];
                    } else if (dp[i - 1][j] < dp[i][j - 1]) {
                        dp[i][j] = dp[i][j - 1];
                        start[i][j] = start[i][j - 1];
                    } else {
                        dp[i][j] = dp[i - 1][j];
                        start[i][j] = Math.max(start[i - 1][j], start[i][j - 1]);
                    }
                }
                
                if (dp[i][j] == T.length()) {
                    if (result == null || i - start[i][j] + 1 < result.length()) {
                        result = S.substring(start[i][j] - 1, i);
                    }
                }
            }
        }
        
        if (result == null) {
            return "";
        }
        return result;
    }
}

File: rust_exercises/leetcode/Problem727/solution.rs
impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let s = s.as_bytes();
        let t = t.as_bytes();
        let mut dp = vec![vec![0 as usize; t.len() + 1]; s.len() + 1];
        let mut start = vec![vec![0 as usize; t.len() + 1]; s.len() + 1];
        let mut result: Option<&[u8]> = None;
        
        for i in 1..s.len() + 1 {
            for j in 1..t.len() + 1 {
                if s[i - 1] == t[j - 1] {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                    if dp[i][j] > 1 {
                        start[i][j] = start[i - 1][j - 1];
                    } else {
                        start[i][j] = i;
                    }
                } else {
                    if dp[i - 1][j] > dp[i][j - 1] {
                        dp[i][j] = dp[i - 1][j];
                        start[i][j] = start[i - 1][j];
                    } else if dp[i - 1][j] < dp[i][j - 1] {
                        dp[i][j] = dp[i][j - 1];
                        start[i][j] = start[i][j - 1];
                    } else {
                        dp[i][j] = dp[i - 1][j];
                        start[i][j] = std::cmp::max(start[i - 1][j], start[i][j - 1]);
                    }
                }
                
                if dp[i][j] == t.len() {
                    result = match result {
                        Some(r) => {
                            if r.len() > i - start[i][j] + 1 {
                                Some(&s[start[i][j] - 1..i])
                            } else {
                                Some(r)
                            }
                        },
                        None => Some(&s[start[i][j] - 1..i]),
                    };
                }
            }
        }
        
        match result {
            Some(r) => r.iter().map(|x| *x as char).collect(),
            None => String::new(),
        }
    }
}
